app/controllers/transactions/transaction-list.controller.it.test.js

```diff
@@ -45,6 +57,30 @@ describe('The /transactions endpoint', () => {
     })
   })
 
+  describe('Error results when from date is later than to date', () => {
+    const request = {
+      account: gatewayAccountFixture.validGatewayAccount({ 'payment_provider': 'stripe' }),
+      service: service,
+      user: user,
+      query: {
+        fromDate: '03/5/2018',
+        fromTime: '01:00:00',
+        toDate: '01/5/2018',
+        toTime: '01:00:00'
+      },
+      url: 'http://selfservice/servce-transactions',
+      session: {}
+    }
+    const response = {
+      render: sinon.spy()
+    }
+    it('should return the response with the date-range failing validation with empty transaction results indicator', async () => {
+      await getController()(request, response, next)
+
+      sinon.assert.calledWith(response.render,'transactions/index',response.render.firstCall.args[1])
```

Comment:
Same as above

Reply:
Yes, its a very verbose object. So I have removed it and the first one will suffice.
The object is below:

```
{
  "total": 0,
  "count": 0,
  "page": 1,
  "results": [],
  "_links": {},
  "filters": {
    "fromDate": "03/5/2018",
    "fromTime": "01:00:00",
    "toDate": "01/5/2018",
    "toTime": "01:00:00"
  },
  "hasFilters": true,
  "hasResults": false,
  "totalOverLimit": false,
  "showCsvDownload": true,
  "totalFormatted": "0",
  "maxLimitFormatted": "5,000",
  "hasPaginationLinks": false,
  "hasPageSizeLinks": false,
  "isInvalidDateRange": true,
  "fromDateParam": "03/5/2018",
  "toDateParam": "01/5/2018",
  "cardBrands": [
    {
      "value": "",
      "text": "All brands",
      "selected": false
    }
  ],
  "downloadTransactionLink": "/account/a-valid-external-id/transactions/download?fromDate=03%2F5%2F2018&toDate=01%2F5%2F2018&fromTime=01%3A00%3A00&toTime=01%3A00%3A00",
  "search_path": "/account/a-valid-external-id/transactions",
  "filtersDescription": " from <strong>03/5/2018</strong> to <strong>01/5/2018</strong>",
  "eventStates": [
    {
      "value": "",
      "text": "Any",
      "selected": false
    },
    {
      "value": "In progress",
      "text": "In progress"
    },
    {
      "value": "Success",
      "text": "Success"
    },
    {
      "value": "Declined",
      "text": "Declined"
    },
    {
      "value": "Timed out",
      "text": "Timed out"
    },
    {
      "value": "Cancelled",
      "text": "Cancelled"
    },
    {
      "value": "Error",
      "text": "Error"
    },
    {
      "value": "Refund submitted",
      "text": "Refund submitted"
    },
    {
      "value": "Refund error",
      "text": "Refund error"
    },
    {
      "value": "Refund success",
      "text": "Refund success"
    },
    {
      "value": "Dispute awaiting evidence",
      "text": "Dispute awaiting evidence"
    },
    {
      "value": "Dispute under review",
      "text": "Dispute under review"
    },
    {
      "value": "Dispute won in your favour",
      "text": "Dispute won in your favour"
    },
    {
      "value": "Dispute lost to customer",
      "text": "Dispute lost to customer"
    }
  ],
  "stateFiltersFriendly": "",
  "clearRedirect": "/account/a-valid-external-id/transactions",
  "isStripeAccount": true,
  "isExperimentalFeaturesEnabled": true,
  "recurringEnabled": false,
  "loggedIn": false,
  "paymentMethod": "card",
  "permissions": {
    "users_service_read": true,
    "users_service_create": true,
    "tokens_active_read": true,
    "tokens_revoked_read": true,
    "tokens_create": true,
    "tokens_update": true,
    "tokens_delete": true,
    "transactions_read": true,
    "transactions_by_date_read": true,
    "transactions_by_fields_read": true,
    "transactions_download_read": true,
    "transactions_details_read": true,
    "transactions_events_read": true,
    "refunds_create": true,
    "transactions_amount_read": true,
    "transactions_description_read": true,
    "transactions_email_read": true,
    "transactions_card_type_read": true,
    "gateway_credentials_read": true,
    "gateway_credentials_update": true,
    "service_name_read": true,
    "service_name_update": true,
    "payment_types_read": true,
    "payment_types_update": true,
    "email_notification_template_read": true,
    "email_notification_paragraph_update": true,
    "email_notification_toggle_update": true,
    "tokens_read": true,
    "toggle_3ds_read": true,
    "toggle_3ds_update": true,
    "users_service_delete": true,
    "merchant_details_read": true,
    "merchant_details_update": true,
    "toggle_billing_address_read": true,
    "toggle_billing_address_update": true,
    "go_live_stage_update": true,
    "go_live_stage_read": true,
    "stripe_account_details_update": true,
    "stripe_bank_details_update": true,
    "stripe_bank_details_read": true,
    "stripe_responsible_person_update": true,
    "stripe_responsible_person_read": true,
    "stripe_director_update": true,
    "stripe_director_read": true,
    "stripe_vat_number_company_number_update": true,
    "stripe_vat_number_company_number_read": true,
    "stripe_government_entity_document_update": true,
    "stripe_organisation_details_update": true,
    "connected_gocardless_account_read": true,
    "connected_gocardless_account_update": true,
    "payouts_read": true,
    "moto_mask_input_update": true,
    "moto_mask_input_read": true,
    "psp_test_account_stage_update": true,
    "webhooks_read": true,
    "webhooks_update": true,
    "agreements_read": true,
    "agreements_update": true
  },
  "isAdminUser": true,
  "hideServiceHeader": false,
  "hideServiceNav": false,
  "hideFeedbackBanner": false,
  "currentGatewayAccount": {
    "payment_provider": "stripe",
    "gateway_account_id": 31,
    "external_id": "a-valid-external-id",
    "allow_apple_pay": false,
    "allow_google_pay": false,
    "service_name": "A fabulous service",
    "type": "test",
    "email_collection_mode": "MANDATORY",
    "email_notifications": {
      "PAYMENT_CONFIRMED": {
        "version": 1,
        "enabled": true,
        "template_body": "template here"
      },
      "REFUND_ISSUED": {
        "version": 1,
        "enabled": true
      }
    },
    "allow_moto": false,
    "moto_mask_card_number_input": false,
    "moto_mask_card_security_code_input": false,
    "recurring_enabled": false,
    "requires3ds": false,
    "integration_version_3ds": 1,
    "disabled": false,
    "full_type": "Stripe test"
  },
  "isTestGateway": true,
  "isSandbox": false,
  "isDigitalWalletSupported": true,
  "currentService": {
    "id": 857,
    "externalId": "cp5wa",
    "name": "System Generated",
    "serviceName": {
      "en": "System Generated"
    },
    "gatewayAccountIds": [
      "666"
    ],
    "collectBillingAddress": false,
    "currentGoLiveStage": "NOT_STARTED",
    "experimentalFeaturesEnabled": true,
    "currentPspTestAccountStage": "NOT_STARTED",
    "agentInitiatedMotoEnabled": false,
    "defaultBillingAddressCountry": "GB",
    "takesPaymentsOverPhone": false
  },
  "humanReadableEnvironment": "Test",
  "serviceNavigationItems": [
    {
      "id": "navigation-menu-home",
      "name": "Dashboard",
      "url": "/account/a-valid-external-id/dashboard",
      "current": false,
      "permissions": true
    },
    {
      "id": "navigation-menu-transactions",
      "name": "Transactions",
      "url": "/account/a-valid-external-id/transactions",
      "current": false,
      "permissions": true
    },
    {
      "id": "navigation-menu-agreements",
      "name": "Agreements",
      "url": "/test/service/:serviceExternalId/account/a-valid-external-id/agreements",
      "current": false,
      "permissions": false
    },
    {
      "id": "navigation-menu-payment-links",
      "name": "Payment links",
      "url": "/account/a-valid-external-id/create-payment-link/manage",
      "current": 0,
      "permissions": true
    },
    {
      "id": "navigation-menu-settings",
      "name": "Settings",
      "url": "/account/a-valid-external-id/settings",
      "current": false,
      "permissions": true
    }
  ],
  "adminNavigationItems": [
    {
      "id": "navigation-menu-settings-home",
      "name": "Settings",
      "url": "/account/a-valid-external-id/settings",
      "current": false,
      "permissions": true
    },
    {
      "id": "navigation-menu-api-keys",
      "name": "API keys",
      "url": "/account/a-valid-external-id/api-keys",
      "current": false,
      "permissions": true
    },
    {
      "id": "navigation-menu-webhooks",
      "name": "Webhooks",
      "url": "/test/service/:serviceExternalId/account/a-valid-external-id/webhooks",
      "current": false,
      "permissions": true
    },
    {
      "id": "navigation-menu-switch-psp",
      "name": "Switch PSP",
      "url": "/account/a-valid-external-id/switch-psp",
      "current": false
    },
    {
      "id": "navigation-menu-payment-types",
      "name": "Card types",
      "url": "/account/a-valid-external-id/payment-types",
      "current": false,
      "permissions": true
    }
  ],
  "_features": {
    "": true
  }
} 
```

I see alot of test use sinon.assert.called(res.render)
I see a lot of test use sinon.assert.called(res.render), so having that improvement in is good.

Reply:
Actually, the following construct works and it so cool.
Thanks Pal.
 ```
sinon.assert.calledWith(res.render, 'api-keys/index', sinon.match({
        'tokens': [],
        'tokens_singular': false
      }))
```